### ISSUE TITLE
Conformance requirements for normative context

### ISSUE THREAD

**Issue by OR13** [w3c/vc-data-model#1185]
`@context` value is now normative.

What should conformant processors expect regarding operation that apply the context.

Should `jsonld.expand` succeed? 

If `jsonld.frame` fails, should a verify raise an error?

etc...

Its not clear why the value is normative, and what part of it, an implementer must understand to conform.

It seems like a few sentences about `frame` or `expand` would be enough to cover this.

**Comment by dlongley** [comment]
We already reference that `@context` is defined by the JSON-LD spec (https://w3c.github.io/vc-data-model/#json-ld) and that spec has its own [conformance section](https://www.w3.org/TR/json-ld11/#conformance). Talking about performing any JSON-LD-related transformations is something that would be done in other specs such as a data integrity cryptosuite spec. So, IMO, I think we're already covered for the core data model spec and can close this.

**Comment by OR13** [comment]
Concrete example, an issuer redefines a protected term in a second context, but doesn't actually apply the context when issuing a credential.

2 verifiers both receive the credential, and verify it successfully but one tries to store it as RDF and encounters the protected term redefined error... The other stores it as JSON.

I guess this highlights that this is a securing spec issue, not a core data model issue?

**Comment by dlongley** [comment]
@OR13,

> Concrete example, an issuer redefines a protected term in a second context, but doesn't actually apply the context when issuing a credential.
> 
> 2 verifiers both receive the credential, and verify it successfully but one tries to store it as RDF and encounters the protected term redefined error... The other stores it as JSON.

In order to [process a VC as "plain" JSON](https://w3c.github.io/vc-data-model/#semantic-interoperability), its contexts must be expected values:

> JSON-based processors MUST process the `@context` key, ensuring the expected values exist in the expected order for the credential type being processed.

This means that the contexts are well-known and have been vetted at least once and either stored permanently or matched up against some expected content hash. It doesn't require that the processing software apply a runtime JSON-LD transform (nor even include a JSON-LD library as a dependency), just that the contexts have been previously vetted (perhaps involving a one-time use of a JSON-LD transform performed by a trusted party) and put into some allow list that the processing software uses.

> I guess this highlights that this is a securing spec issue, not a core data model issue?

It is true that some securing mechanisms will automatically detect problems for unexpected (or unknown) contexts, but for others, if all of the contexts are expected / well-known (and therefore vetted), it isn't a requirement to run any transforms during the consumption or verification process.

**Comment by OR13** [comment]
> it isn't a requirement to run any transforms during the consumption or verification process.

The value it provides only happens when it's processed, and we don't seem to have any requirement for it to ever be processed.

It feels like there should be at least some reason to use the context value, in the core data model.

I don't feel we are giving implementers of the core data model any benefits of JSON-LD over plain JSON.

... But I believe that to be a goal of the working group.

**Comment by dlongley** [comment]
> I don't feel we are giving implementers of the core data model any benefits of JSON-LD over plain JSON.
> 
> ... But I believe that to be a goal of the working group.

I think we are giving them the benefits -- but we're also allowing for interoperability when not all of those benefits are used.

Simpler technologies have fewer requirements for interoperability but also solve fewer use cases and often do not scale. More complexity enables more scaling but at a cost of greater requirements that aren't worth the effort if being part of a large, decentralized ecosystem isn't of interest. A goal of VCs is to build the biggest interoperable, decentralized ecosystem. This means trying to reduce the requirements as much as possible to enable more systems to interop and reuse technology -- even (and especially) when different peers make different use of the available benefits.

**Comment by OR13** [comment]
I appreciate your approval on the PR that defines registered claim names in the V2 context, that is exactly what allowing people to tap the benefits of JSON-LD while not forcing them to process JSON-LD looks like.

**Comment by TallTed** [comment]
>> JSON-based processors MUST process the `@context` key, ensuring the expected values exist in the expected order for the credential type being processed.

> This means that the contexts are well-known and have been vetted at least once and either stored permanently or matched up against some expected content hash.

The quoted spec statement does not require that all `@context` values correspond to previously vetted (in any way) URIs, nor that they be permanently stored nor hash-matched. It only requires that the *expected values exist in the expected order*; it says nothing about unexpected values. If the former is what was intended, we have some (potentially breaking change) rephrasing to do.

**Comment by dlongley** [comment]
@TallTed,

Actually, I believe all of that text is being removed (PR #1182) and could instead end up (with changes) in an implementation guide instead. That would mean that the spec would formally require people to use the JSON-LD conformance section, but the implementation guide could say how that could be done just once by any trusted party to then enable the use of previously vetted, hashed and /or stored, and expected context values in simpler software.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-07-19-vcwg#section2-7) on 2023-07-19

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.7. Conformance requirements for normative context (issue vc-data-model#1185)

_See github issue [vc-data-model#1185](https://github.com/w3c/vc-data-model/issues/1185)._




**Orie Steele:** It is unclear to me whether implementors are ever required to process the context document directly.  

> *Dave Longley:* not sure what Orie wants... but this PR might be related to it or might resolve it?: [https://github.com/w3c/vc-data-model/pull/1202](https://github.com/w3c/vc-data-model/pull/1202).

**Orie Steele:** It would be good to have an explanation of what implementors can rely on now that the context document is required.  

**Manu Sporny:** 1202 was an attempt to explain non-RDF processing.  

> *Orie Steele:* exactly.

> *Orie Steele:* yes! talk about expansion and framing! show the value of JSON-LD... we all the see value of JSON I think.

**Joe Andrieu:** +1 to explaining the value of JSON-LD.  

> *Orie Steele:* +1 need to clarify if the verifier needs to understand normative context.

> *Brent Zundel:* +1 Joe.

> *Manu Sporny:* +1 to JoeAndrieu, agree.

> *Orie Steele:* I think we are on our way to saying that directly... in [https://github.com/w3c/vc-data-model/pull/1202](https://github.com/w3c/vc-data-model/pull/1202).

**Joe Andrieu:** I think that if a party receives a VC, they can assume that their interpretation of the local names is the same.  

---

</details>

**Comment by OR13** [comment]
main open question I have is if verifiers / validators need to understand the context.... in other words... the context is normative, but what actually happens if its "not understood" correctly... I feel our spec should have some text explaining this, given we made it normative.

**Comment by brentzundel** [comment]
It is my understanding that this issue will be addressed by the same PR that addresses #1227

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-09-06-vcwg#section4-5) on 2023-09-06

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 4.5. Conformance requirements for normative context (issue vc-data-model#1185)

_See github issue [vc-data-model#1185](https://github.com/w3c/vc-data-model/issues/1185)._




**Manu Sporny:** oops new comments in the thread, will PR soon.  

**Orie Steele:** related to your last comment on processing the data model in JSON-LD, I think a quick sentence or two would be enough. i'll propose inthread.  

> *Ted Thibodeau Jr.:* also see [https://www.w3.org/TR/string-meta/](https://www.w3.org/TR/string-meta/).

**Joe Andrieu:** I think it might be reductive to say it should never be used in `@context`s, there are places even in VC were it might make sense.  

---

</details>

**Comment by msporny** [comment]
PR #1281 has been raised to address this issue. This issue will be closed once PR #1281 is merged.

**Comment by msporny** [comment]
PR #1281 has been merged, closing.
